src/lib/io/sol-uart-impl-linux.c

```diff
@@ -151,14 +251,35 @@ sol_uart_open(const char *port_name, const struct sol_uart_config *config)
     }
     tcflush(uart->fd, TCIOFLUSH);
 
-    uart->async.rx_fd_handler = sol_fd_add(uart->fd,
-        FD_ERROR_FLAGS | SOL_FD_FLAGS_IN, uart_rx_callback, uart);
-    if (!uart->async.rx_fd_handler) {
-        SOL_ERR("Unable to add file descriptor to watch UART.");
-        goto fail;
+    if (config->rx_cb) {
+        size_t rx_size;
+        void *rx_buf;
+        enum sol_buffer_flags flags = SOL_BUFFER_FLAGS_NO_NUL_BYTE | SOL_BUFFER_FLAGS_DEFAULT;
+
+        uart->fd_handler = sol_fd_add(uart->fd,
+            FD_ERROR_FLAGS | SOL_FD_FLAGS_IN, uart_fd_handler_callback, uart);
+        if (!uart->fd_handler) {
+            SOL_ERR("Unable to add file descriptor to watch UART.");
+            goto fail;
+        }
+
+        uart->rx_cb = config->rx_cb;
+        rx_size = config->rx_size;
+        if (rx_size)
+            flags |= SOL_BUFFER_FLAGS_FIXED_CAPACITY;
+        else
+            rx_size = DEFAULT_BUFFER_SIZE;
+
+        rx_buf = malloc(rx_size);
+        SOL_NULL_CHECK_GOTO(rx_buf, fail);
```

Comment:
i'd move this malloc inside `if (rx_size) { ... }` and on `else { rx_buf = NULL; }`